ocaml/xenopsd/xc/xenops_server_xen.ml

```diff
@@ -3530,47 +3534,70 @@ module VBD = struct
   let vdi_attach_path vbd =
     Printf.sprintf "/xapi/%s/private/vdis/%s" (fst vbd.id) (snd vbd.id)
 
-  let attach_and_activate task xc xs frontend_domid vbd vdi =
+  (* For vdis that are None or local, return attached_vdi, otherwise return the path to attach *)
+  let attached_vdi_from_vdi xs vdi =
+    match vdi with
+    | None ->
+        (* XXX: do something better with CDROMs *)
+        ( Some
+            {
+              domid= this_domid ~xs
+            ; attach_info=
+                Storage_interface.
+                  {
+                    implementations=
+                      [
+                        XenDisk {params= ""; extra= []; backend_type= "vbd3"}
+                      ; BlockDevice {path= ""}
+                      ]
+                  }
+            }
+        , ""
+        )
+    | Some (Local path) ->
+        ( Some
+            {
+              domid= this_domid ~xs
+            ; attach_info=
+                Storage_interface.
+                  {
+                    implementations=
+                      [
+                        XenDisk {params= path; extra= []; backend_type= "vbd3"}
+                      ; BlockDevice {path}
+                      ]
+                  }
+            }
+        , ""
+        )
+    | Some (VDI path) ->
+        (None, path)
+
+  let _attach task xc xs frontend_domid vbd vdi =
```

Comment:
`_` at the beginning is usually reserved for unused functions. If you want an alternative function to avoid name clashes you can use an `'` at the end perhaps?